### PR TITLE
Initial FIELD API and gpu-scc-field variant

### PR DESCRIPTION
**Pull request by mlange05**
_Note: This PR precedes #25 and aims to bring in the base variant of the FIELD API. I'm filing it now as a standalone PR to ease review and integration of changes when we eventually switch to the new long-term supported standalone repo for FIELD API._

This PR revives an old experimentation with optimising data transfer costs via host-side memory pinning behind a common Fortran API for field objects. It therefor ships an early version of the so-called FIELD API (`field_module.F90`) that supports explicit memory pinning for individual field objects, and a bulk-allocation scheme similar to the state array handling in IFS. 

The PR adds a new variant `dwarf-cloudsc-gpu-scc-field` that uses the GPU-SCC kernel for convenience, but allocates field object via the new `CLOUDSC_FIELD_STATE` object. The driver-side FIELD API usage in demonstrated in the new kernel, and the default data transfer times should be superior on Nvidia GPUs to other GPU-SCC variants.

An environment flag to trigger bulk-allocation of arrays via FIELD API wrapper constructors is also provided, and can be triggered via `CLOUDSC_PACKED_STORAGE=ON`, as described in the README. This closely mimics IFS allocation behaviour, but reduces the data transfer performance due to the strided memory layout (thanks @awnawab for showing me this!).

### CODE REVIEW

**Review by reuterbal — APPROVED** [review]
This is great and I very much appreciate this digestable stepping stone towards #25. Very neatly documented and integrated. I've only left a few minor comments, nothing that stops this from going in.

File: arch/toolchains/ecmwf-hpc2020-nvhpc.cmake [hunk 1]
```diff
@@ -50,6 +50,8 @@ set(ECBUILD_Fortran_FLAGS "${ECBUILD_Fortran_FLAGS} -Ktrap=fp")
 set(ECBUILD_Fortran_FLAGS "${ECBUILD_Fortran_FLAGS} -Kieee")
 set(ECBUILD_Fortran_FLAGS "${ECBUILD_Fortran_FLAGS} -Mdaz")
 
+set(ECBUILD_Fortran_LINK_FLAGS "-gpu=pinned")
```
[the remark this answers is not in the document]
Comment (awnawab) [reply arch/toolchains/ecmwf-hpc2020-nvhpc.cmake:53]: Yes I agree, we just need to add this as a link option for the `dwarf-cloudsc-gpu-scc-field` executable, behind a `HAVE_CUDA` test.

**Review by awnawab — APPROVED** [review]
As @reuterbal said, very cleanly done and nicely documented. It also makes the focus of #25 a lot clearer and makes it much easier to integrate; thanks for that! I just mirrored two of B's comments on the cmake side, neither of which is a showstopper by any means, so this can go in as is.

File: src/cloudsc_gpu/dwarf_cloudsc_gpu.F90 [hunk 2]
```diff
@@ -33,17 +33,29 @@ PROGRAM DWARF_CLOUDSC
...
+LOGICAL :: USE_PACKED
 
 INTEGER(KIND=JPIM) :: NUMOMP   = 1     ! Number of OpenMP threads for this run
 INTEGER(KIND=JPIM) :: NGPTOTG  = 16384 ! Number of grid points (as read from command line)
 INTEGER(KIND=JPIM) :: NPROMA   = 64   ! NPROMA blocking factor (currently active)
 INTEGER(KIND=JPIM) :: NGPTOT           ! Local number of grid points
 
+
+#ifdef CLOUDSC_GPU_SCC_FIELD
+TYPE(CLOUDSC_FIELD_STATE) :: FIELD_STATE
```
[the remark this answers is not in the document]
Comment (mlange05) [reply src/cloudsc_gpu/dwarf_cloudsc_gpu.F90]: Yes, although FIELD_STATE has an additional option. However, the call to verify now aliases nicely, and I moved the invocation to the back to keep with the overall order.

**Comment by mlange05** [comment]
Ok, thanks for the reviews. I think I've addressed all point, but please check. I've also had to rebase, but the changes since last review are in the last 3 commits.

**Comment by reuterbal** [comment]
Looks all good now, many thanks! GTG

**Pull request merged**